feat: oso `workflow` improvements + `gemini` tutorial setup

Jabolol opened the pull request:
> [!NOTE]
> Description in progress, WIP.

**Closes:** #4807, #4521, #4767, #4715

---

## Issue #4807: "Phoenix experiments get overridden"

**Problem:** Only the last experiment run gets saved, losing past runs.

**How the PR addresses it:**

**File:** [`warehouse/oso_agent/oso_agent/datasets/uploader.py`](warehouse/oso_agent/oso_agent/datasets/uploader.py)

**Changes:** Enhanced dataset upload logic with comprehensive logging and better handling of existing vs new datasets

### Key improvements:
- Added detailed logging throughout upload process to track what's happening
- Better differentiation between subset mode (which deletes existing) vs full mode (which appends)
- Clear warnings when destructive operations occur: 
  ```
  DESTRUCTIVE OPERATION: Running in subset mode...
  ```
- Improved dataset comparison logic to avoid overriding existing experiments

---

## Issue #4521: "Add agent experiment model annotations"

**Problem:** Need metadata about experiments (git hash, workflow info, LLM model, etc.)

**How the PR addresses it:**

**File:** [`warehouse/oso_agent/oso_agent/util/metadata.py`](warehouse/oso_agent/oso_agent/util/metadata.py)

**Changes:** Created new module for experiment metadata collection

### Key improvements:
- `get_git_commit_hash()` - captures current git commit
- `get_git_branch()` - captures current branch
- `get_git_dirty_status()` - checks for uncommitted changes
- `get_workflow_info()` - captures workflow class details
- `get_llm_info()` - captures LLM configuration
- `collect_experiment_metadata()` - combines all metadata

**File:** [`warehouse/oso_agent/oso_agent/eval/text2sql/experiment.py`](warehouse/oso_agent/oso_agent/eval/text2sql/experiment.py)

**Changes:** Updated experiment runners to use the new metadata collection:

```python
experiment_metadata = collect_experiment_metadata(
    config=config,
    workflow_cls=BasicText2SQL,
    additional_metadata={"name": config.agent_name},
)
```

---

## Issue #4767: "Connect workflows to vector database"

**Problem:** Workflows need access to vector database context.

**How the PR addresses it:**

**File:** [`warehouse/oso_agent/oso_agent/workflows/text2sql/mixins.py`](warehouse/oso_agent/oso_agent/workflows/text2sql/mixins.py)

**Changes:** Split `OsoQueryEngineWorkflowMixin` into `OsoVectorDatabaseMixin` (base functionality) and `OsoQueryEngineWorkflowMixin` (complete implementation)
- This separation allows workflows to use vector database functionality independently

**File:** [`warehouse/oso_agent/oso_agent/workflows/text2sql/semantic.py`](warehouse/oso_agent/oso_agent/workflows/text2sql/semantic.py)

**Changes:** Enhanced semantic workflow to use vector database context
- Added `entity_context` parameter to semantic translation
- Connected vector retrievers to provide context for better query translation
- Improved retry logic with vector context

**File:** [`warehouse/oso_agent/oso_agent/tool/oso_semantic_query_tool.py`](warehouse/oso_agent/oso_agent/tool/oso_semantic_query_tool.py)

**Changes:** Added entity context support:

```python
# Entity Context (relevant entities from vector database)
{entity_context}
```

**File:** [`warehouse/oso_agent/oso_agent/workflows/default.py`](warehouse/oso_agent/oso_agent/workflows/default.py)

**Changes:** Enhanced workflow resolver to provide vector database resources:

```python
embedding = resources.get_resource("embedding")
storage_context = resources.get_resource("storage_context")
```

---

## Issue #4715: "End to end Colab tutorial setup"

**Problem:** Need complete tutorial from setup to notebook generation.

**How the PR addresses it:**

**File:** [`apps/docs/docs/guides/notebooks/gemini.md`](apps/docs/docs/guides/notebooks/gemini.md)

**Changes:** Created comprehensive tutorial documentation

### Key improvements:
- **Installation:** Step-by-step Gemini CLI installation
- **Authentication:** Google OAuth and API key setup
- **OSO MCP Setup:** Docker configuration and environment variables
- **Testing:** Connection verification steps
- **Notebook Generation:** Complete example with World Chain analysis
- **Troubleshooting:** Common issues and solutions

**File:** [`warehouse/oso_agent/oso_agent/server/app.py`](warehouse/oso_agent/oso_agent/server/app.py)

**Changes:** Simplified the text2sql endpoint to focus on semantic workflow
- Removed dual workflow execution (basic + semantic)
- Streamlined to use only semantic workflow for better performance
- This supports the tutorial's goal of providing a clean, focused experience

---

## Workflow Diagram

```mermaid
graph TD
    A[Text2SQLStartEvent] --> B[initiate_query_engine]
    B --> C[StartQueryEngineEvent]

    C --> D[OsoVectorDatabaseMixin:<br/>analyze_schema]
    D --> E[SchemaAnalysisEvent]

    E --> F[OsoVectorDatabaseMixin:<br/>retrieve_row_context]
    F --> G[RowContextEvent<br/>+ vector retrievers]

    G --> H[SemanticText2SQLWorkflow:<br/>handle_row_context_for_semantic_translation]
    H --> H1[Extract entity context<br/>from vector retrievers]
    H1 --> H2[Call _perform_semantic_translation<br/>max_retries + empty error_context]

    H2 --> I[SemanticText2SQLWorkflow:<br/>_perform_semantic_translation]
    I --> I1[semantic_query_tool.acall<br/>with entity_context]
    I1 --> J{Success?}

    J -->|Yes| K[SemanticQueryEvent]
    J -->|No| L[RetrySemanticQueryEvent]

    L --> M[SemanticText2SQLWorkflow:<br/>start_translation retry handler]
    M --> M1[Get vector context directly<br/>for retry attempt]
    M1 --> M2[Call _perform_semantic_translation<br/>with retry state]
    M2 --> I

    K --> N[SemanticText2SQLWorkflow:<br/>build_sql_from_semantic_query]
    N --> O{Success?}
    O -->|Yes| P[Text2SQLGenerationEvent]
    O -->|No| L

    P --> Q[GenericText2SQLRouter:<br/>handle_text2sql_generation_event]
    Q --> R{execute_sql?}

    R -->|Yes| S[SQLExecutionRequestEvent]
    R -->|No| T[StopEvent<br/>SqlResponse]

    S --> U[OsoDBWorkflow:<br/>retrieve_sql_results]
    U --> V[SQLResultEvent]

    V --> W[GenericText2SQLRouter:<br/>handle_sql_results_rows]
    W --> WW{synthesize_response?}
    WW -->|Yes| X[SQLResultSummaryRequestEvent]
    WW -->|No| Y[StopEvent<br/>AnyResponse]

    X --> Z[SQLRowsResponseSynthesisMixin:<br/>synthesize_sql_response_from_rows]
    Z --> AA[SQLResultSummaryResponseEvent]

    AA --> BB[SemanticText2SQLWorkflow:<br/>return_response]
    BB --> CC[StopEvent<br/>StrResponse]

    %% OsoVectorDatabaseMixin - Light Blue
    style D fill:#b3e5fc,color:#000
    style F fill:#b3e5fc,color:#000

    %% SemanticText2SQLWorkflow - Light Purple
    style H fill:#e1bee7,color:#000
    style I fill:#e1bee7,color:#000
    style M fill:#e1bee7,color:#000
    style N fill:#e1bee7,color:#000
    style BB fill:#e1bee7,color:#000

    %% Vector Context Operations - Light Green
    style H1 fill:#c8e6c9,color:#000
    style M1 fill:#c8e6c9,color:#000

    %% Enhanced Semantic Tool - Yellow
    style I1 fill:#ffcc02,color:#000

    %% GenericText2SQLRouter - Light Orange
    style Q fill:#ffcc80,color:#000
    style W fill:#ffcc80,color:#000

    %% OsoDBWorkflow - Light Red
    style U fill:#ffcdd2,color:#000

    %% SQLRowsResponseSynthesisMixin - Light Cyan
    style Z fill:#b2dfdb,color:#000
```

ravenac95 commented on warehouse/oso_agent/oso_agent/util/metadata.py:
```diff
@@ -0,0 +1,179 @@
+
+
+def get_system_info() -> dict[str, t.Any]:
+    """Get system information.
```
👍

ravenac95 reviewed (changes requested):
Cool mostly looks good just need to address a couple things I noted with regards to the metadata.

Jabolol commented:
> Cool mostly looks good just need to address a couple things I noted with regards to the metadata.

Addressed your feedback, thanks for taking a look :)

ravenac95 reviewed (approved):
Looks good!
